File: libs/python/stonecharts/charts/xrange.py
"""X-range / Gantt chart renderer: ChartSpec -> SVG string.

Horizontal span bars on lane categories with optional milestones and
dependency connectors. Rides the shared cartesian frame with
orientation="horizontal", x_scale="band", include_zero=False.
"""

from __future__ import annotations

from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian


def render_svg(spec) -> str:
    lane_cats = spec.y_axis.categories or []
    lane_count = len(lane_cats)
    spec.x_axis.categories = lane_cats

    for s in spec.series:
        while len(s.data) < lane_count:
            s.data.append(0.0)

    all_times: list[float] = []
    for s in spec.series:
        for sp in s.spans or []:
            all_times.append(sp.x)
            all_times.append(sp.x2)
    if spec.y_axis.min is None:
        spec.y_axis.min = spec.x_axis.min if spec.x_axis.min is not None else (min(all_times) if all_times else 0.0)
    if spec.y_axis.max is None:
        spec.y_axis.max = spec.x_axis.max if spec.x_axis.max is not None else (max(all_times) if all_times else 0.0)

    return render_cartesian(spec, "X-range", "band", _xrange_marks, include_zero=False, orientation="horizontal")


PAD = 0.2


def _xrange_marks(fr: CartesianFrame, p: list[str]) -> None:
    if fr.n <= 0:
        return

    lane_height = fr.plot_h / fr.n
    bar_thickness = lane_height * (1 - PAD)

    def ypix_band(j: int) -> float:
        return fr.plot_y + lane_height * j + lane_height / 2

    def xval(v: float) -> float:
        return fr.value_pix(v)

    cats = fr.cats

    span_index: dict[str, tuple[float, float]] = {}
    for si, s in enumerate(fr.spec.series):
        for sp in s.spans or []:
            if sp.id:
                end_x = xval(sp.x2)
                cy = ypix_band(sp.y)
                span_index[sp.id] = (end_x, cy)

    for si, s in enumerate(fr.spec.series):
        st = fr.styles[si]
        fill = st.fill

        p.append(f'<g class="sc-series" data-series="{si}">')

        for sp in s.spans or []:
            start = sp.x
            end = sp.x2
            lane = sp.y

            x_left = xval(min(start, end))
            x_right = xval(max(start, end))
            w = x_right - x_left
            cy = ypix_band(lane)
            cx = xval((start + end) / 2)

            lane_label = cats[lane] if lane < len(cats) else str(lane)
            duration = end - start

            data_attrs = (
                f'data-series="{si}" data-series-name="{esc(s.name)}" '
                f'data-x="{esc(lane_label)}" data-y="{esc(fmt_num(start))}" '
                f'data-start="{esc(fmt_num(start))}" data-end="{esc(fmt_num(end))}" '
                f'data-lane="{esc(lane_label)}" data-duration="{esc(fmt_num(duration))}" '
                f'data-color="{st.solid}" data-r="{fmt_num(3.5)}" data-r-hover="{fmt_num(6)}" '
                f'cx="{cx:.1f}" cy="{cy:.1f}"'
            )

            if sp.milestone:
                h = bar_thickness / 2
                points = f"{cx:.1f},{cy - h:.1f} {cx + h:.1f},{cy:.1f} {cx:.1f},{cy + h:.1f} {cx - h:.1f},{cy:.1f}"
                p.append(f'<polygon class="sc-milestone sc-point" {data_attrs} points="{points}" fill="{fill}"/>')
            else:
                if w < 1.0:
                    w = 1.0
                top = cy - bar_thickness / 2
                p.append(
                    f'<rect class="sc-span sc-point" {data_attrs} '
                    f'x="{x_left:.1f}" y="{top:.1f}" width="{w:.1f}" height="{bar_thickness:.1f}" fill="{fill}"/>'
                )

            if sp.dependency:
                for dep_id in sp.dependency:
                    if dep_id in span_index:
                        pred_x, pred_y = span_index[dep_id]
                        this_x = xval(start)
                        this_y = cy
                        mid_x = (pred_x + this_x) / 2
                        ah = 4.0
                        d = (
                            f"M{pred_x:.1f} {pred_y:.1f} "
                            f"L{mid_x:.1f} {pred_y:.1f} "
                            f"L{mid_x:.1f} {this_y:.1f} "
                            f"L{this_x:.1f} {this_y:.1f} "
                            f"M{this_x - ah:.1f} {this_y - ah:.1f} "
                            f"L{this_x:.1f} {this_y:.1f} "
                            f"L{this_x - ah:.1f} {this_y + ah:.1f}"
                        )
                        p.append(
                            f'<path class="sc-dependency" d="{d}" '
                            f'fill="none" stroke="{st.solid}" stroke-width="1" opacity="0.5"/>'
                        )

        p.append("</g>")

File: libs/go/xrange.go
package stonecharts

import (
	"fmt"
	"math"
	"strings"
)

func renderXRangeSVG(spec *ChartSpec) string {
	laneCats := spec.YAxis.Categories
	laneCount := len(laneCats)
	spec.XAxis.Categories = laneCats

	for i := range spec.Series {
		s := &spec.Series[i]
		for len(s.Data) < laneCount {
			s.Data = append(s.Data, 0.0)
		}
	}

	allTimes := []float64{}
	for _, s := range spec.Series {
		for _, sp := range s.Spans {
			allTimes = append(allTimes, sp.X, sp.X2)
		}
	}
	if spec.YAxis.Min == nil && len(allTimes) > 0 {
		if spec.XAxis.Min != nil {
			v := *spec.XAxis.Min
			spec.YAxis.Min = &v
		} else {
			v := allTimes[0]
			for _, t := range allTimes[1:] {
				if t < v {
					v = t
				}
			}
			spec.YAxis.Min = &v
		}
	}
	if spec.YAxis.Max == nil && len(allTimes) > 0 {
		if spec.XAxis.Max != nil {
			v := *spec.XAxis.Max
			spec.YAxis.Max = &v
		} else {
			v := allTimes[0]
			for _, t := range allTimes[1:] {
				if t > v {
					v = t
				}
			}
			spec.YAxis.Max = &v
		}
	}

	return renderCartesian(spec, "X-range", "band", xrangeMarks, false, "horizontal")
}

const xrPAD = 0.2

func xrangeMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	laneHeight := f.plotH / float64(f.n)
	barThickness := laneHeight * (1 - xrPAD)

	ypixBand := func(j int) float64 {
		return f.plotY + laneHeight*float64(j) + laneHeight/2
	}

	xval := func(v float64) float64 {
		return f.valuePix(v)
	}

	cats := f.cats

	type spanRef struct {
		endX float64
		cy   float64
	}
	spanIndex := map[string]spanRef{}
	for _, s := range f.spec.Series {
		for _, sp := range s.Spans {
			if sp.ID != "" {
				spanIndex[sp.ID] = spanRef{xval(sp.X2), ypixBand(sp.Y)}
			}
		}
	}

	for si, s := range f.spec.Series {
		st := f.styles[si]
		fill := st.fill

		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))

		for _, sp := range s.Spans {
			start := sp.X
			end := sp.X2
			lane := sp.Y

			xLeft := xval(math.Min(start, end))
			xRight := xval(math.Max(start, end))
			w := xRight - xLeft
			cy := ypixBand(lane)
			cx := xval((start + end) / 2)

			laneLabel := fmt.Sprintf("%d", lane)
			if lane < len(cats) {
				laneLabel = cats[lane]
			}
			duration := end - start

			dataAttrs := fmt.Sprintf(
				`data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-start="%s" data-end="%s" data-lane="%s" data-duration="%s" data-color="%s" data-r="%s" data-r-hover="%s" cx="%s" cy="%s"`,
				si, esc(s.Name), esc(laneLabel), esc(fmtNum(start)),
				esc(fmtNum(start)), esc(fmtNum(end)),
				esc(laneLabel), esc(fmtNum(duration)),
				st.solid, fmtNum(3.5), fmtNum(6),
				f1(cx), f1(cy))

			if sp.Milestone {
				h := barThickness / 2
				points := fmt.Sprintf("%s,%s %s,%s %s,%s %s,%s",
					f1(cx), f1(cy-h), f1(cx+h), f1(cy), f1(cx), f1(cy+h), f1(cx-h), f1(cy))
				p.WriteString(fmt.Sprintf(`<polygon class="sc-milestone sc-point" %s points="%s" fill="%s"/>`, dataAttrs, points, fill))
			} else {
				if w < 1.0 {
					w = 1.0
				}
				top := cy - barThickness/2
				p.WriteString(fmt.Sprintf(`<rect class="sc-span sc-point" %s x="%s" y="%s" width="%s" height="%s" fill="%s"/>`,
					dataAttrs, f1(xLeft), f1(top), f1(w), f1(barThickness), fill))
			}

			for _, depID := range sp.Dependency {
				if pred, ok := spanIndex[depID]; ok {
					predX := pred.endX
					predY := pred.cy
					thisX := xval(start)
					thisY := cy
					midX := (predX + thisX) / 2
					ah := 4.0
					d := fmt.Sprintf("M%s %s L%s %s L%s %s L%s %s M%s %s L%s %s L%s %s",
						f1(predX), f1(predY), f1(midX), f1(predY),
						f1(midX), f1(thisY), f1(thisX), f1(thisY),
						f1(thisX-ah), f1(thisY-ah), f1(thisX), f1(thisY),
						f1(thisX-ah), f1(thisY+ah))
					p.WriteString(fmt.Sprintf(`<path class="sc-dependency" d="%s" fill="none" stroke="%s" stroke-width="1" opacity="0.5"/>`,
						d, st.solid))
				}
			}
		}

		p.WriteString(`</g>`)
	}
}
